packages/ai-jsx/src/core/conversation.tsx

```diff
@@ -0,0 +1,295 @@
+    element.tag === AssistantMessage ||
+    element.tag === SystemMessage ||
+    element.tag === FunctionCall ||
+    element.tag === FunctionResponse
+  );
+}
+
+function toConversationMessages(partialRendering: AI.PartiallyRendered[]): ConversationMessage[] {
+  return partialRendering
+    .flatMap((e) => (AI.isElement(e) ? [e] : []))
+    .map<ConversationMessage>((e) => {
+      switch (e.tag) {
+        case UserMessage:
+          return { type: 'user', element: e };
+        case AssistantMessage:
+          return { type: 'assistant', element: e };
+        case SystemMessage:
+          return { type: 'system', element: e };
+        case FunctionCall:
+          return { type: 'functionCall', element: e };
+        case FunctionResponse:
+          return { type: 'functionResponse', element: e };
+        default:
+          throw new AIJSXError(
+            `Unexpected tag (${e.tag.name}) in conversation`,
+            ErrorCode.UnexpectedPartialRenderResult,
+            'internal'
+          );
+      }
+    });
+}
+
+/** @hidden */
+export async function renderToConversation(
+  conversation: AI.Node,
+  render: AI.ComponentContext['render']
+): Promise<ConversationMessage[]> {
+  return toConversationMessages(await render(conversation, { stop: isConversationalComponent }));
+}
+
+/**
+ * A component that appends messages to a conversation according to a `reply` function.
+ *
+ * The `reply` function is invoked with the new messages produced from its _own replies_ until
+ * it fails to return a conversational message (e.g. UserMessage or AssistantMessage). The first
+ * invocation includes the messages from the `children` prop.
+ *
+ * @example
+ * ```tsx
+ *    <Converse reply={(messages) => messages[0].type === "user" ? <AssistantMessage>Hello there!</AssistantMessage> : null}>
+ *      <UserMessage>Hello!</UserMessage>
+ *    </ChatCompletion>
+ *
+ *    ==> 'Hello there!'
+ * ```
+ *
+ */
+export async function* Converse(
+  {
+    reply,
+    children,
+  }: {
+    reply: (messages: ConversationMessage[], fullConversation: ConversationMessage[]) => AI.Renderable;
+    children: AI.Node;
+  },
+  { render, memo }: AI.ComponentContext
+): AI.RenderableStream {
+  yield AI.AppendOnlyStream;
+
+  let fullConversation = [] as ConversationMessage[];
+  let next = memo(children);
+  while (true) {
+    const messages = await renderToConversation(next, render);
+    if (messages.length === 0) {
+      break;
+    }
+
+    fullConversation = fullConversation.concat(messages);
+    next = memo(reply(messages, fullConversation));
+    yield next;
+  }
+
+  return AI.AppendOnlyStream;
+}
+
+/**
+ * Allows the presentation of conversational components ({@link UserMessage} et al) to be altered.
+ *
+ * Also accepts an `onComplete` prop, which will be invoked once per render with the entire conversation.
+ *
+ * @example
+ * ```tsx
+ *     <ShowConversation present={(msg) => msg.type === "assistant" && <>Assistant: {msg.element}</>}>
```

Comment:
Hmm, I'm not fully convinced that `present` belongs at this layer, as opposed to just giving a full JSON to the caller, and letting the UI decide what to show.

I guess there's a case where we might not want to return function call/responses to the client, and this is an easier place to enforce that than modifying the final output stream itself?

Reply:
Yeah the tricky part (as usual) is streaming (and/or potentially embedded UI), which is why these transformations are in tree-space rather than JSON- or text- space.

Reply:
Per our conversation going to leave this for now. As our approach to UI/RSC/embedded React/etc. crystalizes I'm totally open to reworking this.